### ISSUE TITLE
Controlled input props value

### ISSUE THREAD

**Issue by r0kan** [s-yadav/react-number-format#749]
#### Describe the issue and the actual behavior
The value of the input should not change if the value in the props has not changed

#### Describe the expected behavior
Field value changes

#### Provide a CodeSandbox link illustrating the issue
https://codesandbox.io/s/controlled-input-ybgr6z

Is this expected behaviour ? How to reset field value to value in props ?

**Comment by s-yadav** [comment]
In v5 the component doesn't block editing, behaviour came from v4. As this would be a breaking change its planned for next major release.
Meanwhile, you can create your custom controlled wrapper on top of it.
https://github.com/s-yadav/react-number-format/issues/700#issuecomment-1344398443

**Comment by azevedo** [comment]
It seems not possible to change the input's value from the outside, except on the first render, while on v4 it was. `value` acts as `defaultValue` it seems.

The example below was based on one of your examples. I've added some buttons that change the value passed to the `value` prop, but nothing changes. The only way to change is by typing in the input.

https://codesandbox.io/s/custom-numeric-format-forked-s1qtpj

**Comment by s-yadav** [comment]
@azevedo looks like this is working fine on nonstrict mode. But breaking on strict mode. Will check on why it's breaking on strict mode.
Here's a non strict mode example.
https://codesandbox.io/s/custom-numeric-format-forked-7l5ti9?file=/src/App.js

**Comment by crcorbett** [comment]
Hi,

Just letting you know I'm also having difficulty with `value` failing to update to match changes in the `value` prop after the first render. My use case is I provide `<NumericFormat />` with defaultValues from React Hook Form, which are updated once the component is attached to the client using values sourced from Local Storage.

Debugging with React Dev Tools I notice that the `value` prop within `NumericFormat` and `NumberFormatBase` are out-of-sync. `NumericFormat` appears to have the correct (updated) value, whereas `NumberFormatBase` has the previous default value.

If useful I can try and set up a code sandbox to example this issue for you.

Importantly, and as you have noted, `value` does work correctly when React is in non-strict mode. This leads me to believe this might just be a simple bug caused by a rogue ref / useEffect somewhere in `NumberFormatBase`.

**Comment by s-yadav** [comment]
@crcorbett this is fixed on 5.2.2. Let me know if you are still facing any issue.

**Comment by crcorbett** [comment]
Amazing @s-yadav!

Can confirm the `values` prop now updates as expected. Thank you so much for your swift fix, your hard work is greatly appreciated!

**Comment by TibiTotoro** [comment]
Hi, it still doesn't work for me on 5.2.2.

**Comment by kytosai** [comment]
Hi, it still doesn't work for me on 5.3.1

**Comment by s-yadav** [comment]
The controlled component wrapper example not working for you. 
https://codesandbox.io/s/controlled-numeric-format-w236xh?file=/src/App.js

**Comment by fxckdead** [comment]
It's not working on `5.3.1`.

@s-yadav on your controlled snippet it's not really working as a controlled input because it should not update the input value if you don't update the `propValue` or `localValue`, but it still allows you to type data:

```jsx
function ControlledNumericFormat({
  value: propValue,
  onValueChange,
  ...props
}) {
  const [localValue, setLocalValue] = React.useState(propValue);
  useEffect(() => {
    if (propValue !== localValue) {
      setLocalValue(propValue);
    }
  }, [propValue, localValue]);

  function handleChange(values) {
    onValueChange(values);
    // setLocalValue(values.floatValue); <-- this commented shouldn't allow input changes
  }

  return (
    <NumericFormat value={localValue} onValueChange={handleChange} {...props} />
  );
}
```

**Comment by ambrosil** [comment]
i can confirm too it's not working on 5.3.1

**Comment by kubek93** [comment]
I would ask you to change the priority of this issue. It's a very common case when components are needed to change the value from outside. I'm doing it using react-query on filters clear. I was trying to fix the issue long hours, so everyone who is trying to fix the issue here is solutions:

```ts
<Controller
        key={inputKey + fromTo}
        name={`${inputKey}.${fromTo}`}
        control={control}
        render={({ field }) => {
          return (
            <NumericFormat
              value={field.value}
              displayType={"text"}
              decimalSeparator: ",",
              decimalScale: 2,
              renderText={(value) => {
                const parsedValue = field.value ? value : "";

                return (
                  <Input
                    {...field}
                    value={parsedValue}
                    size="sm"
                    onBlur={(event) => {
                      const value = inputNumberOnBlur(event);
                      field.onChange(value);
                      field.onBlur();
                    }}
                  />
                );
              }}
            />
          );
        }}
/>
```

Code is listening if `react-hook-form` state has been changed and if `field.value` is undefined we have to set up empty string.

**Comment by ambrosil** [comment]
yes i confirm it was a normal rerendering problem, related to react rerendering and not related to react-number-format.
thanks everybody.

**Comment by blagojablazhevski** [comment]
> I would like to ask you to prioritize this issue. It's a very common scenario where we need to change the value from the outside. I'm currently addressing it using `react-hook-form` when filters are cleared. After spending long hours attempting to fix the issue, here are the solutions that everyone who is trying to fix it should consider:
> 
> ```ts
> <Controller
>   key={inputKey + fromTo}
>   name={`${inputKey}.${fromTo}`}
>   control={control}
>   render={({ field }) => {
>     return (
>       <NumericFormat
>         value={field.value}
>         displayType={"text"}
>         decimalSeparator: ",",
>         decimalScale: 2,
>         renderText={(value) => {
>           const parsedValue = field.value ? value : "";
> 
>           return (
>             <Input
>               {...field}
>               value={parsedValue}
>               size="sm"
>               onBlur={(event) => {
>                 const value = inputNumberOnBlur(event);
>                 field.onChange(value);
>                 field.onBlur();
>               }}
>             />
>           );
>         }}
>       />
>     );
>   }}
> />
> ```
> 
> The code listens for changes in the `react-hook-form` state, and if `field.value` is `undefined`, we need to set it to an empty string to avoid react warnings.

This definitely needs to be high priority, it's a shame it still hasn't been fixed since it's such a common use case. I'm also having the same issue with setting the value past the first render with react-hook-form.

**Comment by Erwol** [comment]
Hey, I came across this issue while trying to set min and max values (this other issue: https://github.com/s-yadav/react-number-format/issues/380). I created a workaround that works when using `react-number-format` with MUI's `TextField`; but you can probably adapt it to whatever input field you are using. Any feedback is welcome: https://codesandbox.io/p/sandbox/ml9phj

**Comment by rtatarinov** [comment]
Any updates here?

**Comment by willy2dg** [comment]
upvoting this, i have the same case like @erwol

**Comment by willy2dg** [comment]
After some analisis i found a solution. The input is controlled, but the 'onChange' on TextField is fired before 'onValueChange'. Then what happens when you modify the value in onValueChange instead of just updating ?

If you want a max value of 7, and you type 9, your logic changes and updates the state to 7, it refresh the state and the TextField is updated, it works fine the first time. But if you type 9 again, the TextField is updated to 79, your logic gives you 7 again, but the current state ( value state ), still 7, so it isn't resfreshed and the TextField keeps his internal value and seems uncontrolled. But the problem is the state value never changed,

So whats the solution ? If you want to alter the onValueChange result, you should reply the same logic on 'isAllowed' to avoid changes on the TextField when the value state will remain the same, so we will disallow changes giving the same state result.

```
value={value}
isAllowed={values => {
  let targetValue = values.floatValue
  const isNumber = typeof targetValue == 'number'
  if (isNumber && maxValue) targetValue = Math.min(targetValue, maxValue)
  if (isNumber && minValue) targetValue = Math.max(targetValue, minValue)
  return value !== targetValue
}}
onValueChange={(values, { source }) => {
  let targetValue = values.floatValue
  const isNumber = typeof targetValue == 'number'
  if (source == 'event' && typeof onChange == 'function') {
    if (isNumber && maxValue) targetValue = Math.min(targetValue, maxValue)
    if (isNumber && minValue) targetValue = Math.max(targetValue, minValue)
    setValue(isNumber ? targetValue : '')
  }
}}
```

in other words 

```
value={value}
isAllowed={values => {
  let targetValue = computeFinalValue(values)
  return value !== targetValue
}}
onValueChange={values => {
  let targetValue = computeFinalValue(values)
  setValue(isNumber ? targetValue : '')
}}
```

maybe there is some mistake while adapting the example, but the concept works
